Source/discord/discord.cpp

```diff
@@ -0,0 +1,176 @@
+#include "discord.h"
+
+#include <Discord/cpp/discord.h>
+
+#include <algorithm>
+#include <array>
+#include <cctype>
+#include <chrono>
+#include <random>
+#include <sstream>
+#include <string>
+#include <tuple>
+
+#include <fmt/format.h>
+
+#include "../gendung.h"
+#include "../panels/charpanel.hpp"
+#include "../player.h"
+#include "../utils/language.h"
+
+// App ID used when someone launches classic Diablo
+constexpr discord::ClientId DiscordDiabloAppId = 496571953147150354;
+
+// App ID used for DevilutionX's Diablo
+constexpr discord::ClientId DiscordDevilutionxAppId = 795760213524742205;
+
+namespace devilution {
+
+namespace {
+constexpr std::array<const char *, 3> DifficultyStrs = { N_("Normal"), N_("Nightmare"), N_("Hell") };
+constexpr std::array<const char *, DTYPE_LAST + 1> DungeonStrs = { N_("Town"), N_("Cathedral"), N_("Catacombs"), N_("Caves"), N_("Hell"), N_("Nest"), N_("Crypt") };
+
+constexpr auto IgnoreResult = [](discord::Result result) {};
+} // namespace
+
+DiscordManager::DiscordManager()
+{
+	discord::Result result = discord::Core::Create(DiscordDevilutionxAppId, DiscordCreateFlags_NoRequireDiscord, &core_);
+	good_ = result == discord::Result::Ok;
+}
+
+DiscordManager &DiscordManager::Instance()
+{
+	static DiscordManager instance {};
+	return instance;
+}
+
+std::string DiscordManager::GetLocationString() const
+{
+	const char *dungeonStr = _(/* TRANSLATORS: type of dungeon (i.e. Cathedral, Caves)*/ "None");
+	if (dungeon_area_ != DTYPE_NONE) {
+		dungeonStr = _(DungeonStrs.at(dungeon_area_));

```

Comment:
`std::out_of_range` is never caught, might as well use `operator[]` in this case. Also given `dungeon_level_` is an enum type using explicit casts here will handle potential future changes to scoped enums.

```suggestion
		dungeonStr = _(DungeonStrs[static_cast<std::size_t>(dungeon_area_)]);
```

This suggestion can be applied to the other uses of `at()` with no exception handling, and the uses of enum values with arithmetic operations.

Reply:
I'd rather have `at()` throw an uncaught exception that the unhandled exception filter would then report on. `operator[]` can create hard to diagnose bugs by allowing buffer overflows without incident. For example if someone is modding or the capability to add more dungeon types becomes natively supported through json or something, but this particular file is forgotten. Given that kind of risk and no drawback I don't see a benefit to using `operator[]` over `at()` in this piece.

Reply:
Also I think global exception filters are only available on Windows?

Reply:
> Also I think global exception filters are only available on Windows?

There's a standard library equivalent but if everyone favours `operator[]` I'll change it.